### ISSUE TITLE
Passing arguments to nested queries

### ISSUE THREAD

**Issue by mishaszu** [async-graphql/async-graphql#1337]
Hello!

I'm looking to switch from `juniper` to `async-graphql`. I'm giving it a try and I like it for the most part, but I found myself struggling with one thing.
So far I found that context in `async-graphql` is more like a global thing while in `juniper` I construct it per request. It is particularly useful when I want to pass an argument to all nested queries.
For example, if I want to pass the user id or time range to all nested object resolvers I can just write to context.
I can't figure out a good and safe way to pass an argument to nested resolvers in `async-graphql`. The only workaround I can think of is passing an argument explicitly to all fields... but it's kind of tedious. I wonder if there is a better way.
I would appreciate help with finding a solution for the problem I encounter!

**Comment by sunli829** [comment]
You can attach context data when constructing the schema:

https://docs.rs/async-graphql/6.0.1/async_graphql/struct.SchemaBuilder.html#method.data

Or attach context data on every request:

https://docs.rs/async-graphql/6.0.1/async_graphql/struct.Request.html#method.data

Then you can get it in field resolver:

https://docs.rs/async-graphql/6.0.1/async_graphql/context/struct.ContextBase.html#method.data

**Comment by mishaszu** [comment]
I found [an example in repo](https://github.com/async-graphql/examples/blob/b433457d3bea5ce91b4ad63aaef769206a6ca814/poem/token-from-header/src/main.rs#L39) showing how to set request `data`.

Adding mutable data to schema doesn't make much sense in my example because I don't want to leak data between requests.
Attaching context data to request would probably work and I could use data in a request scope, but it would probably add extra burden to what I would like to achieve.
I want to pass query variables and attach them to context from the resolver level. It would be convenient because in a resolver I can access deserialized input objects. At least in `juniper` case where it's possible to mutate context from resolver.

I want to contain data within GrapqhQL request, I don't want to add extra params to post request. Following the documentation I found it's possible to get query variables from GraphQLRequest in the form of [async_graphql::Value](https://docs.rs/async-graphql/6.0.1/async_graphql/enum.Value.html). Maybe attaching it to request data is a better way, as mutating context from resolvers might be tricky.

Summing up I have two questions remaining:
- is there any utility helping to transform [async_graphql::value](https://docs.rs/async-graphql/6.0.1/async_graphql/enum.Value.html) into actual graphql input object?
- is it safe to assume attaching things to request data won't leak between requests?

**Comment by sunli829** [comment]
https://github.com/async-graphql/examples/tree/master/poem/token-from-header

I think this example can help you 😄

**Comment by mishaszu** [comment]
Oh! You're right! I followed how the token is obtained from the header and tried similar for variables. From the documentation, I totally missed that `async_graphql::Value` `to_json` returns a Result with `serde_json::value::Value` and not `async_graphql::Value`. It all totally makes sense now, as I can deserialize with `serde_json` 😄 

Thank you a lot for your help, I think it answers all my questions!
